src/utils/pnp.py

```diff
@@ -0,0 +1,313 @@
+    -------
+    q : ndarray of shape (4,)
+        The unit quaternion (w, x, y, z)
+    """
+    q = np.empty([4, ])
+
+    tr = np.trace(R)
+    if tr < 0:
+        i = R.diagonal().argmax()
+        j = (i + 1) % 3
+        k = (j + 1) % 3
+
+        q[i] = np.sqrt(1 - tr + 2 * R[i, i]) / 2
+        q[j] = (R[j, i] + R[i, j]) / (4 * q[i])
+        q[k] = (R[k, i] + R[i, k]) / (4 * q[i])
+        q[3] = (R[k, j] - R[j, k]) / (4 * q[i])
+    else:
+        q[3] = np.sqrt(1 + tr) / 2
+        q[0] = (R[2, 1] - R[1, 2]) / (4 * q[3])
+        q[1] = (R[0, 2] - R[2, 0]) / (4 * q[3])
+        q[2] = (R[1, 0] - R[0, 1]) / (4 * q[3])
+
+    q /= np.linalg.norm(q)
+    # Rearrange (x, y, z, w) to (w, x, y, z)
+    q = q[[3, 0, 1, 2]]
+
+    return q
+
+
+def Quaternion2Rotation(q):
+    """
+    Convert a quaternion to rotation matrix
+
+    Parameters
+    ----------
+    q : ndarray of shape (4,)
+        Unit quaternion (w, x, y, z)
+    Returns
+    -------
+    R : ndarray of shape (3, 3)
+        The rotation matrix
+    """
+    q /= np.linalg.norm(q)
+
+    w = q[0]
+    x = q[1]
+    y = q[2]
+    z = q[3]
+
+    R = np.empty([3, 3])
+    R[0, 0] = 1 - 2 * y ** 2 - 2 * z ** 2
+    R[0, 1] = 2 * (x * y - z * w)
+    R[0, 2] = 2 * (x * z + y * w)
+
+    R[1, 0] = 2 * (x * y + z * w)
+    R[1, 1] = 1 - 2 * x ** 2 - 2 * z ** 2
+    R[1, 2] = 2 * (y * z - x * w)
+
+    R[2, 0] = 2 * (x * z - y * w)
+    R[2, 1] = 2 * (y * z + x * w)
+    R[2, 2] = 1 - 2 * x ** 2 - 2 * y ** 2
+
+    return R
+
+def skewsymm(x):
+    Sx = np.zeros((3, 3))
+    Sx[0, 1] = -x[2]
+    Sx[0, 2] = x[1]
+    Sx[1, 0] = x[2]
+    Sx[2, 0] = -x[1]
+    Sx[1, 2] = -x[0]
+    Sx[2, 1] = x[0]
+
+    return Sx
+
+
+def SetupPnPNL(C_T_G, G_p_f, C_b_f):
+    n_points = G_p_f.shape[1]
+    n_projs = n_points
+    b = np.zeros((3 * n_projs,))
+
+    for k in range(n_points):
+        b[3 * k: 3 * (k + 1)] = C_b_f[:, k]
+
+    z = np.zeros((7 + 3 * n_points,))
+    R = C_T_G[:3, :3]
+    t = C_T_G[:3, 3]
+    q = Rotation2Quaternion(R)
+
+    z[0:7] = np.concatenate([t, q])
+    for i in range(n_points):
+        z[7 + 3 * i: 7 + 3 * (i + 1)] = G_p_f[:, i]
+
+    return z, b
+
+
+def MeasureReprojectionSinglePose(z, b, p, w):
+
+    n_projs = b.shape[0] // 3
+    f = np.zeros((3 * n_projs,))
+    s = 0.001 * np.ones((3 * n_projs,))
+
+    q = z[3:7]
+    q_norm = np.sqrt(np.sum(q ** 2))
+    q = q / q_norm
+    R = Quaternion2Rotation(q)
+    t = z[:3]
+
+    for j in range(n_projs):
+        X = p[3 * j:3 * (j + 1)]
+        # Remove measurement error of fixed poses
+        b_hat = R @ X + t
+        f[3 * j: 3 * (j + 1)] = b_hat / np.sqrt(np.sum(b_hat ** 2))
+        s[3 * j: 3 * (j + 1)] = w[j]
+
+    err = s * (b - f)
+
+    return err
+
+
+def UpdatePose(z):
+
+    p = z[0:7]
+    q = p[3:]
+
+    q = q / np.linalg.norm(q)
+    R = Quaternion2Rotation(q)
+    t = p[:3]
+    P_new = np.hstack([R, t[:, np.newaxis]])
+
+    return P_new
+
+
+def P3PKe(m, X, inlier_thres=1e-5):
```

Comment:
Is this P3P the approach proposed in [Ke and Roumeliotis 2017]. If so, add a comment